### DISCUSSION TITLE
Utilizing Symfony Secret Vault

### DISCUSSION THREAD

**Discussion by mc-supply** [pimcore/pimcore#7725] [Q&A / Support]
I'm attempting to move towards more secure storage of sensitive information.  Looking at the ability to secure this information using the built-in Symfony "Secret Vault" and "bind" parameters, similar to this: https://symfonycasts.com/screencast/symfony5-upgrade/secrets-vault-setup

Question I have is, what is the best way to to access these variables from Pimcore Models?  My use case is I make API calls to populate some of the values, and I'd like to have credentials/keys/etc encrypted.  I have thought about doing it using Events, but the way I designed mine was to have a generic ObjectListener:

    <?php

    namespace AppBundle\EventListener;
  
    use Pimcore\Event\Model\ElementEventInterface;

    class ObjectListener
    {
        public function onPreAdd(ElementEventInterface $e)
        {
            $object = $e->getObject();
        
            if (method_exists($object, 'onPreAdd')) {
                $object->onPreAdd();
            }
        }
    
        public function onPreUpdate(ElementEventInterface $e)
        {
            $object = $e->getObject();
        
            if (method_exists($object, 'onPreUpdate')) {
                $object->onPreUpdate();
            }
        }
    
        public function onPostUpdate(ElementEventInterface $e)
        {
            $object = $e->getObject();
        
            if (method_exists($object, 'onPostUpdate')) {
                $object->onPostUpdate();
            }
        }
    }

This makes it a bit tougher to inject "bind" variables, since the parameter(s) could vary, depending on the DataObject.  Is there a better approach, short of creating a new EventListener for each DataObject?

**Comment by passioneight** [comment]
A little late, but may be helpful anyway.

The task you are facing is indeed a bit tricky, as you'll always have to use `instanceof` to check for the data objects type. Otherwise you don't have a way to know in which context you are calling the corresponding method on your object.

> As you stated, a possible solution is to handle each data object in their own listener.

#### Updating your code
One way to improve your code is to use an `interface` and check if the object implements it - this way you don't need to use `method_exists` and also can leverage auto-completion. However, it does not resolve the initial issue: knowing the context.

The good news is, you already implemented the `onPreAdd`, `onPreUpdate`, `onPostUpdate` methods in your objects. Thus, within your object, you are already aware of the context. So, you could just pass an `array` with _all parameters_, independent of the data object (but still checking for your `interface`). Then, the data objects themselves will only use the values they need.

> Checking for your `interface` will ensure that the methods are not called just because they exist. Someone else might have implemented a method with the same name, but a different purpose.

#### Using another approach
You could also just add the [encrypted fields](https://pimcore.com/docs/pimcore/current/Development_Documentation/Objects/Object_Classes/Data_Types/Others.html#page_Encrypted-Field) (mind the selected version of the documentation) to your class definition. Then, in your listener (which really should be a subscriber), you can call `$object->setValues($params)` with `$params` holding any value you want to set - e.g..

```php
$params = [
    'apiKey' => 'whateverValueYouNeedForYourKey',
    ...
];
```

> Whenever your `$object` contains a `set<paramName>` method (e.g., `setApiKey`), the method is called and the value is set on the object. Note, though, that this is potentially dangerous.

#### Additional notes
Personally, I always try to avoid listeners, as my colleagues might not know much about the project, thus, having a hard time realizing that something is happening within a listener. Additionally, your implementation will eventually affect the performance for the corresponding actions.

Maybe you can use a queue instead and then do any performance-heavy tasks in the background, when processing the queue. 

> This depends on whether you need to process the information right away.

Surely, someone else will have yet another way to achieve what you are trying to do (or have arguments against using any of the described approaches). In the end it all comes down to what's right for _your project_. Thus, none of us will ever have enough information about the project (unless participating), so, we can only offer _possible_ soultions.
